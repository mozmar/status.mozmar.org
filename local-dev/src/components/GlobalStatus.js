import React, { Component } from 'react';

import favFailed from '../img/favicon-failed.png';
import favHealthy from '../img/favicon-healthy.png';
import favWarning from '../img/favicon-warning.png';

class GlobalStatus extends Component {
    static propTypes = {
        clearDesktopNotify: React.PropTypes.func.isRequired,
        desktopNotify: React.PropTypes.bool.isRequired,
        message: React.PropTypes.string.isRequired,
        notifyMessage: React.PropTypes.bool.isRequired,
        status: React.PropTypes.string.isRequired
    }

    statusToColor(status) {
        var result;

        switch(status) {
            case 'healthy':
                result = 'success';
                break;
            case 'warning':
                result = 'warning';
                break;
            case 'failed':
                result = 'danger';
                break;
            case 'pending':
                result = 'info';
                break;
            default:
                result = '';
                break;
        }

        return result;
    }

    statusToIcon(status) {
        var img;

        switch (status) {
            case 'healthy':
                img = favHealthy;
                break;
            case 'pending':
            case 'warning':
                img = favWarning;
                break;
            case 'failed':
                img = favFailed;
                break;
            default:
                break;
        }

        return img;
    }

    setFavicon(status) {
        var favicon = document.getElementById('favicon');
        var img = this.statusToIcon(status);

        if (favicon && img) {
            favicon.href = img;
        }
    }

    sendDesktopNotification(message, icon) {
        var options = {
            body: 'Mozilla Engagement Engineering Status Board',
            icon: icon
        };

        new Notification(message, options);

        // Make sure notification isn't duplicated if view changes before next
        // global status request.
        this.props.clearDesktopNotify();
    }

    render() {
        this.setFavicon(this.props.status);

        var className = 'alert alert-' + this.statusToColor(this.props.status);

        if (this.props.desktopNotify === true && this.props.notifyMessage === true) {
            this.sendDesktopNotification(this.props.message, this.statusToIcon(this.props.status));
        }

        return (
            <div className={className} role="alert">
                {this.props.message}
            </div>
        );
    }
}

export default GlobalStatus;
