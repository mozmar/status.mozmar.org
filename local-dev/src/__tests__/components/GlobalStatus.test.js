import React from 'react';
import renderer from 'react-test-renderer';

import GlobalStatus from '../../components/GlobalStatus';

const mockClearDesktopNotify = jest.fn();

describe('render GlobalStatus', () => {
  test('renders a healthy GlobalStatus', () => {
      const component = renderer.create(
          <GlobalStatus status='healthy'
                        message='test'
                        notifyMessage={false}
                        desktopNotify={false}
                        clearDesktopNotify={mockClearDesktopNotify} />
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
  });

  test('renders a warning GlobalStatus', () => {
      const component = renderer.create(
          <GlobalStatus status='healthy'
                        message='test'
                        notifyMessage={false}
                        desktopNotify={false}
                        clearDesktopNotify={mockClearDesktopNotify} />
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
  });

  test('renders a failed GlobalStatus', () => {
      const component = renderer.create(
          <GlobalStatus status='healthy'
                        message='test'
                        notifyMessage={false}
                        desktopNotify={false}
                        clearDesktopNotify={mockClearDesktopNotify} />
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
  });

  test('renders a pending GlobalStatus', () => {
      const component = renderer.create(
          <GlobalStatus status='healthy'
                        message='test'
                        notifyMessage={false}
                        desktopNotify={false}
                        clearDesktopNotify={mockClearDesktopNotify} />
      );

      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
  });
});

test('GlobalStatus sends a desktop notification', () => {
    // tests can't actually send a desktop notification
    GlobalStatus.prototype.sendDesktopNotification = jest.fn(() => true);
    // tests don't like looking for file paths, so we create a predictable stub
    GlobalStatus.prototype.statusToIcon = jest.fn(() => 'stub');

    const component = renderer.create(
        <GlobalStatus status='healthy'
                      message='playtime is fun'
                      notifyMessage={true}
                      desktopNotify={true}
                      clearDesktopNotify={mockClearDesktopNotify} />
    );

    expect(GlobalStatus.prototype.sendDesktopNotification).toHaveBeenCalledWith('playtime is fun', 'stub');
});
