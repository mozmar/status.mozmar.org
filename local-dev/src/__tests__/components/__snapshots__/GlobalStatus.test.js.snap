exports[`render GlobalStatus renders a failed GlobalStatus 1`] = `
<div
  className="alert alert-success"
  role="alert">
  test
</div>
`;

exports[`render GlobalStatus renders a healthy GlobalStatus 1`] = `
<div
  className="alert alert-success"
  role="alert">
  test
</div>
`;

exports[`render GlobalStatus renders a pending GlobalStatus 1`] = `
<div
  className="alert alert-success"
  role="alert">
  test
</div>
`;

exports[`render GlobalStatus renders a warning GlobalStatus 1`] = `
<div
  className="alert alert-success"
  role="alert">
  test
</div>
`;
